File: 6. Zigzag Conversion/6. Zigzag Conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        rows = [[] for _ in range(numRows)]
        pos = 0
        d_pos = 1

        for c in s:
            rows[pos].append(c)

            if pos == numRows - 1:
                d_pos = -1
            elif pos == 0:
                d_pos = 1

            pos += d_pos

        for i in range(numRows):
            rows[i] = "".join(rows[i])

        return "".join(rows)

File: 6. Zigzag Conversion/6. Zigzag Conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
  if (numRows === 1) {
    return s;
  }

  const rows = Array.from({ length: numRows }, () => []);
  let pos = 0;
  let dPos = 1;

  for (const c of s) {
    rows[pos].push(c);

    if (pos === numRows - 1) {
      dPos = -1;
    } else if (pos === 0) {
      dPos = 1;
    }

    pos += dPos;
  }

  for (let i = 0; i < numRows; i++) {
    rows[i] = rows[i].join("");
  }

  return rows.join("");
};
